### DISCUSSION TITLE
H4 not keeping sync with spindle

### DISCUSSION THREAD

**Discussion by Kisssys** [kachurovskiy/nanoels#120] [Q&A] [answered]
Hello Maxim,

I received my H4 control a few days ago and have it installed and mostly working.  All seems to be well, the direction, start stop everything seems to be working fine.

When I hand rotate the spindle the degree readout reads correctly.  It gives reasonably correct rpm when the spindle is running.

Unfortunately if I run the spindle even at low speed and bring it to a stop it has lost the orignianl position of the spindle.  I have a 1024 E6B2 Omron encoder that was working fine with a Clough42.  I've looked at the A B signals and the 5 volt line with a scope and they seem to be acceptable with only a little noise on the highs and lows but with a good fast transition. I've powered the encoder from the H4 and from a separate 5 volt supply with no difference.

I tried to cut threads and of course it isn't tracking.  I have a 600 P/R encoder coming today but hate to change it as my Clough42 is set up and working and easy to put back.  I am running is slow enough that the pulse rate even with a 1024 encoder should be fine.

Maybe I'm assuming that it always tracks the spindle and my procedure  is faulty.

I had 2 handwheels hooked to Z and X and they worked fine but I went back and turned them off to see if that changed anything and it didn't.  

Any ideas

Cheers
Kisssys

**Comment by kachurovskiy** [comment]
Hi @Kisssys, please upload the latest h4.ino to your H4 and let me know if it helps. There was a bug in the original source code version. Cheers!

**Reply by Kisssys** [reply]
Ok, I'll give it a try!

**Reply by Kisssys** [reply, marked as answer]
This is all installed on a Jet BD920 lathe.

Well I think I'm all good now but had a few bumps in the road.  Take my comments for what they are worth, just my experience in getting it going.

I installed the new code and did get a good thread cut so I hope that solves that.

Since I ordered the 600 P/R encoder I decided to install it so that we are comparing apples to apples if I have further problems.  It reads 1/1 with the spindle.

After putting it in I lost the angle all together.  It turns out the encoder I bought off of Amazon (Taiss/Incremental Rotary Encoder DC 5-24v)is voltage sensitive and if the voltage fed to it is less than 4.7 volts it quits.  That brought be to question of how to solve the problem of the H4 losing several hundred milivolts on the 5volt line.  Seems as though the fuses and filters add about 20 ohms in series with the 5 volt line.  I bit the bullet and shorted out both fuses and the 200 ohm resistors and I have a much more stable system.  200 ohms on the A1 and A2 protects those lines but makes them useless to power a rotary encoder which draws about 70 ma each.

With 2 rotary encoders my total current is about 250 ma.  You can find the encoder on Printables at 
https://www.printables.com/model/497597-handwheel-pulse-encoder-dual-axis

My Leadshine ES-D508 on Z and DM556T on X both have limiting resistors of about 270 ohms in their optocoupler circuit. The addition of 200 ohms in series is unnecessary and detrimental.  Most circuits to stepper controllers expect a 0 ohm resistor in series if the driving voltage is 5 volts.

All the protections are good but I think they might cause you grief in the long run. 

I use 5 volt power supplys that are approved for the Raspberry Pi and have a C type connector. I've found a lot of 5 volt supplys are only good to charge your phone with.

I'm sure I'll have a wish list down the road but the only thing interface wise I find unusual is how the X axis when 0 goes away on the screen.  Seems Like Z and X when showing, both should be their.

This is perfect for the type of lathe work I do.  The Clough42 worked perfectly but you had to open and close the half nut to start and stop which made it of limited value to me. My Jet BD920 I find hard to drop in accurately with a .0625 inch leadscrew.

Now I need to go use it for a few weeks.

Thanks for a great product.  

Cheers
Kisssys

**Reply by kachurovskiy** [reply]
Glad that the spindle encoder problem is resolved.

As per EN 61010-1 -+10% voltage fluctuations should not cause problems, it's unfortunate that the encoder was so sensitive.

Your MPG model for is very nice, thank you for sharing. It's a very nice idea to use the Ethernet cable there.

As you noted, A1 and A2 are not designed to be used for connecting encoders and the fact that it works for some encoder models is a happy coincidence. A better approach there would be to only use the I/O pins of A1 and A2 and take 5V & GND from an external power source.

I'm seriously considering to drop the 200 Ohm resistors on the terminals and the fuse for the rotary encoder thanks to your feedback.

To show X position even when it's 0, change https://github.com/kachurovskiy/nanoels/blob/main/h4/h4.ino#L640 `} else if (xDisplayPos != 0) {` to `} else {`.

> This is all installed on a Jet BD920 lathe.

I would be very thankful if you could share a couple pictures and a few words about how you use H4 in https://github.com/kachurovskiy/nanoels/discussions/new?category=show-and-tell - thank you!

**Reply by Kisssys** [reply]
All sounds good.  I'll try the code changes and I'll be happy to do a little show and tell.  I uploaded my encoder mount and x axis mounts to printables and thingiverse.

Cheers
Steve 
Kisssys

**Reply by Kisssys** [reply]
I just noticed beep() in your code and I've never beeped on anything.  Don't really care but I thought I'd mention it.

I've use all modes several times and getting the hang of it.  Need to play with CONE more.

I added the code change and its better, but I may have caused other formating issues as shown in the uploaded photos. I have seen it cut off the letter P in position and add a quote symbol but couldn't reproduce it this morning.  Would just the letters Z and X be as descriptive as printing Position?

```
if (lcdHashLine2 != newHashLine2) {
    lcdHashLine2 = newHashLine2;
    charIndex = 0;
    lcd.setCursor(0, 2);
    if (z.isManuallyDisabled) {
      charIndex += lcd.print("Z disable");
    } else {
      if (xDisplayPos == 0 && !x.isManuallyDisabled) {
        charIndex += lcd.print("Position ");
      }
      charIndex += printAxisPos(&z);
    }
    while (charIndex < 10) {
      charIndex += lcd.print(" ");
    }
    if (x.isManuallyDisabled) {
      charIndex += lcd.print("X disable");
    } else  {   //} else if (xDisplayPos != 0) {
      charIndex += printAxisPos(&x);
       }
    printLcdSpaces(charIndex);
  }
```
I just noticed beep() in your code and I've never beeped on anything.  Don't really care but I thought I'd mention it.



![IMG_1184](https://github.com/kachurovskiy/nanoels/assets/5122480/8669a36d-82ee-4cfc-b611-23f54d211735)
![IMG_1185](https://github.com/kachurovskiy/nanoels/assets/5122480/155ea01b-0dc3-4141-92b9-8922744418fe)
![IMG_1186](https://github.com/kachurovskiy/nanoels/assets/5122480/68559046-3e93-4f35-856c-dc57f8b762b6)

Cheers
Steve

**Reply by kachurovskiy** [reply]
> I just noticed beep() in your code and I've never beeped on anything. Don't really care but I thought I'd mention it.

In the first few devices it's pretty quiet because I over-designed that circuit and added R7 resistor inline with the buzzer. You can hear it if you try to start e.g. thread mode without specifying the stops. To fix the issue, replace that resistor with a bead of solder - apologies for the trouble.

![image](https://github.com/kachurovskiy/nanoels/assets/517919/418e2e36-a089-4c57-bef1-b66fb32b404e)

>  Would just the letters Z and X be as descriptive as printing Position?

Sure, you can adjust your code as you see fit - changing "Position " string to something else or deleting these 3 lines entirely:

```
      if (xDisplayPos == 0 && !x.isManuallyDisabled) {
        charIndex += lcd.print("Position ");
      }
```

**Comment by Kisssys** [comment]
Ok Sounds good
Cheers
Steve
